xtra-libp2p/src/endpoint.rs

```diff
@@ -530,26 +530,40 @@ impl Endpoint {
     async fn handle(
         &mut self,
         msg: OpenSubstream<Single>,
-        _: &mut Context<Self>,
+        ctx: &mut Context<Self>,
     ) -> Result<Pin<Box<dyn futures::Future<Output = Result<Substream, Error>> + Send>>, Error>
     {
-        let peer = msg.peer_id;
+        let peer_id = msg.peer_id;
         let protocols = msg.protocols;
 
         debug_assert!(
             protocols.len() == 1,
             "Type-system enforces that we only try to negotiate one protocol"
         );
 
-        let (control, _) = self.controls.get(&peer).ok_or(Error::NoConnection(peer))?;
+        let (control, _) = self
+            .controls
+            .get(&peer_id)
+            .ok_or(Error::NoConnection(peer_id))?;
 
+        let this = ctx.address().expect("self to be alive");
         let fut = {
             let connection_timeout = self.connection_timeout;
             let control = control.clone();
             async move {
-                let (protocol, stream) =
-                    Self::open_substream(control, peer, protocols.clone(), connection_timeout)
-                        .await?;
+                let res =
+                    Self::open_substream(control, peer_id, protocols.clone(), connection_timeout)
+                        .await;
+
+                if let Err(Error::BadConnection(e)) = &res {
+                    tracing::debug!(
+                        %peer_id,
+                        "Disconnecting peer due to yamux connection error when opening substream: {e}"
+                    );
+                    let _ = this.send_async_safe(Disconnect(peer_id)).await;
```

Comment:
Should this not be a send_async_next (I forget the exact name)

Reply:
I like your first suggestion, primarily because then we don't have to ignore the result.

Do you think it is important that we give this message priority though? I suppose the earlier we get it processed the quicker we'll stop using the faulty connection. But it didn't strike me as something that needed to be done immediately.

Reply:
I think it makes sense to give priority in case the next queued message was a Ping message or similar - in that case we will skip the message sending to the disconnected peer